CLN: Deduplicate show_versions

TomAugspurger opened the pull request:
Closes #26814 

The old version had special cases for numpy, scipy, and pytz. All those now provide a standard `__version__`. xlrd and xlwt are the only special ones now (`__version__`)

I messed up the minimum xlwt version in the last PR. I copied openpyxl's rather than xlwt's. That's now fixed.


I prettied up the output format a bit, to align the version numbers. Hope that 's OK.

```
INSTALLED VERSIONS
------------------
commit           : cbc6edf8bfad33c8bc698a3d035b593d921b0112
python           : 3.7.3.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.6.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.0.dev0+729.gcbc6edf8b
pytest           : 4.6.2
pip              : 19.1.1
setuptools       : 41.0.1
Cython           : 0.29.10
numpy            : 1.17.0.dev0+5b06588
scipy            : 1.3.0
pyarrow          : 0.13.0
xarray           : 0.12.1
IPython          : 7.5.0
sphinx           : 1.8.5
patsy            : 0.5.1
dateutil         : 2.7.3
pytz             : 2018.5
blosc            : 1.5.1
bottleneck       : 1.2.1
tables           : 3.5.1
numexpr          : 2.6.8
feather          : 0.4.0
matplotlib       : 3.1.0
openpyxl         : 2.5.11
xlrd             : 1.1.0
xlwt             : 1.3.0
xlsxwriter       : 1.1.2
lxml.etree       : 4.2.5
bs4              : 4.7.1
html5lib         : 1.0.1
sqlalchemy       : 1.2.14
pymysql          : 0.9.2
psycopg2         : 2.7.6.1 (dt dec pq3 ext lo64)
jinja2           : 2.10.1
s3fs             : 0.1.6
fastparquet      : None
pandas_gbq       : 0.10.0
pandas_datareader: None
gcsfs            : 0.1.2

```

jreback commented on pandas/util/_print_versions.py:
```diff
@@ -60,56 +61,54 @@ def show_versions(as_json=False):
+        "IPython",
+        "sphinx",
+        "patsy",
+        "dateutil",
```
can you only list the ones that are *not* in pandas.compat._optional (and just import from that list)?

jreback commented on pandas/util/_print_versions.py:
```diff
@@ -60,56 +61,54 @@ def show_versions(as_json=False):
+        "xlrd",
+        "xlwt",
+        "xlsxwriter",
+        "lxml.etree",
```
lxml is also a weird one, is it still like this?

TomAugspurger replied:
Ahh I missed that. Yea it needs to be fixed.

TomAugspurger replied:
Err wait, we're fine as long as the `modname` we pass is `lxml.etree`. So I think we're OK.

jreback commented:
formatting looks nice!

TomAugspurger commented:
Got a bit carried away here, sorry. 

1. Fixed xlwt version in the docs / VERSIONS
2. Updated pytables to use the new way, added to docs / VERSIONS
3. Added a test decorator for skipping if a library is installed
4. Deduplicate show_versions (the intent of this PR).

updated output:

```

INSTALLED VERSIONS
------------------
commit           : a8e0c2dff540ceada75fec07c6ccd4e6f0e79039
python           : 3.7.3.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.6.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.0.dev0+730.ga8e0c2dff.dirty
numpy            : 1.17.0.dev0+5b06588
pytz             : 2018.5
dateutil         : 2.7.3
pip              : 19.1.1
setuptools       : 41.0.1
Cython           : 0.29.10
pytest           : 4.6.2
hypothesis       : 3.59.1
sphinx           : 1.8.5
blosc            : 1.5.1
feather          : 0.4.0
xlsxwriter       : 1.1.2
lxml.etree       : 4.2.5
html5lib         : 1.0.1
pymysql          : 0.9.2
psycopg2         : 2.7.6.1 (dt dec pq3 ext lo64)
jinja2           : 2.10.1
IPython          : 7.5.0
pandas_datareader: None
bs4              : 4.7.1
bottleneck       : 1.2.1
fastparquet      : None
gcsfs            : 0.1.2
matplotlib       : 3.1.0
numexpr          : 2.6.8
openpyxl         : 2.5.11
pandas_gbq       : 0.10.0
pyarrow          : 0.13.0
tables           : 3.5.2
s3fs             : 0.1.6
scipy            : 1.3.0
sqlalchemy       : 1.2.14
xarray           : 0.12.1
xlrd             : 1.1.0
xlwt             : 1.3.0
```

jreback commented on pandas/util/_test_decorators.py:
```diff
@@ -140,6 +140,17 @@ def skip_if_no(
     )
 
 
+def skip_if_installed(
```
are u using this?

TomAugspurger replied:
https://github.com/pandas-dev/pandas/pull/26816/files#diff-56c44e2fe6ac59a333e18f097db07524R57.

We should probably have a basic test like this for all our optional deps. Thoughts?

It can't go in `test_pytables.py` since that does an `importorskip` at the top of the file.

TomAugspurger replied:
Alternatively, we could *try* monkeypatching / mocking the module, but I've never had luck working around Python's import caching.

jreback replied:
not sure i understand? we already test all of the optional depa but using them
are you suggesting something else?

TomAugspurger replied:
> are you suggesting something else?

Yeah, testing when that *aren't* present.

simonjayhawkins commented on pandas/tests/test_optional_dependency.py:
```diff
@@ -50,3 +52,11 @@ def test_no_version_raises():
         import_optional_dependency(name)
+
+
+@td.skip_if_installed("tables")
```
in the interest of better coverage, would it not be better to monkeypatch `builtins.__import__` as was done in #26685 for tests like this rather than skip?

gfyoung commented on pandas/compat/_optional.py:
```diff
@@ -16,13 +16,13 @@
     "xarray": "0.8.2",
     "xlrd": "1.0.0",
-    "xlwt": "2.4.0",
+    "xlwt": "1.0.0",
```
Why is this being downgraded?

TomAugspurger replied:
I messed up in the previous PR. That was openpyxl's min version.

TomAugspurger commented:
I messed up in the previous PR. That was openpyxl's min version.

On Wed, Jun 12, 2019 at 3:52 PM gfyoung <notifications@github.com> wrote:

> *@gfyoung* commented on this pull request.
> ------------------------------
>
> In pandas/compat/_optional.py
> <https://github.com/pandas-dev/pandas/pull/26816#discussion_r293112317>:
>
> >      "s3fs": "0.0.8",
>      "scipy": "0.19.0",
>      "sqlalchemy": "1.1.4",
>      "xarray": "0.8.2",
>      "xlrd": "1.0.0",
> -    "xlwt": "2.4.0",
> +    "xlwt": "1.0.0",
>
> Why is this being downgraded?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/26816?email_source=notifications&email_token=AAKAOITCRUVDRB35ZWRRICLP2FOXNA5CNFSM4HXPCCTKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3LZ4FY#pullrequestreview-249011735>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIW33LL7I2P3I6KJEXTP2FOXNANCNFSM4HXPCCTA>
> .
>

jreback commented:
lgtm. merge on green.

TomAugspurger committed:
update

TomAugspurger commented:
Apologies for the force push.

I've given up on my `missing_dependency` fixture. Split that off to https://github.com/pandas-dev/pandas/issues/26902.

jreback commented on pandas/tests/test_optional_dependency.py:
```diff
@@ -50,3 +52,11 @@ def test_no_version_raises():
+
+@td.skip_if_installed("tables")
+def test_pytables_raises():
+    df = pd.DataFrame({"A": [1, 2]})
```
can you go ahead and create

pandas/tests/io/pytables/test_missing_deps.py and put this there (we need to split test_pytables anyhow)

jreback reviewed (changes requested):
IIRC this had a lot more changes (meaning this actually did the de-duplication of show_versions)

TomAugspurger committed:
restore cleanup

TomAugspurger committed:
Merge remote-tracking branch 'upstream/master' into 26814-optional-fu

TomAugspurger committed:
move test

TomAugspurger commented:
> IIRC this had a lot more changes (meaning this actually did the de-duplication of show_versions)

Sorry, that de-duplication is restored now.

jreback commented:
thanks @TomAugspurger very nice!
